### PR TITLE
chore(docker): bump python-train base to CUDA 12.6.3 + cuDNN 9

### PR DESCRIPTION
**Pull request by ayutaz**
## Summary

Trivy code-scanning が `piper-plus-python-train` Docker image に対して **3,469 件の OS-package CVE** を検出している (libssl / glibc / libxml2 / openssl 等)。 これは base image を 2023-04 リリースの `nvidia/cuda:12.1.0-cudnn8-devel-ubuntu22.04` に固定したまま 2 年以上経過し、 apt layer が老朽化した結果。

Base image を **CUDA 12.6.3 + cuDNN 9 + Ubuntu 22.04** にアップグレードし、 OS layer の CVE を一括解消する。 PyTorch / Lightning / piper_train の API 互換性に影響しない範囲の保守的な bump に絞る。

## Affected Components

- [ ] Python runtime (`src/python_run/`)
- [ ] C# runtime (`src/csharp/`)
- [ ] Rust runtime (`src/rust/`)
- [ ] C++ runtime (`src/cpp/`)
- [ ] Go runtime (`src/go/`)
- [ ] WASM / npm runtime (`src/wasm/`)
- [x] Docker (`docker/`)
- [ ] CI / CD (`.github/`)
- [ ] Documentation (`docs/`, `README.md`)

## Type

- [ ] Bug fix
- [ ] New feature
- [ ] Refactoring
- [ ] Documentation
- [ ] CI / CD
- [x] Dependencies

## Risk Level

- [ ] patch (内部実装 / バグ修正 / 依存パッチ更新 — 動作変化なし)
- [x] minor (新機能追加 — 既存 API 互換)
- [ ] major (破壊的変更 — API / 動作変化)

## Contract Impact

- [x] None (`docs/spec/*.toml` 影響なし — Docker base image のみ更新)

## 変更内容

| 機能 / 対象 | 動作 | これがないと起こること |
|------------|------|----------------------|
| `docker/python-train/Dockerfile` builder stage | `nvidia/cuda:12.1.0-cudnn8-devel-ubuntu22.04` → `nvidia/cuda:12.6.3-cudnn-devel-ubuntu22.04` | builder layer の apt-installed packages (build-essential / libssl / glibc 等) が古いまま、 Trivy CVE が継続検出される |
| `docker/python-train/Dockerfile` runtime stage | `nvidia/cuda:12.1.0-cudnn8-runtime-ubuntu22.04` → `nvidia/cuda:12.6.3-cudnn-runtime-ubuntu22.04` | runtime layer の apt-installed packages が古いまま、 配布される image の CVE が解消されない |

## 設計判断

- **cuDNN 8 → 9 (image side) は実行時影響なし**: PyTorch 2.2.1+cu121 wheel は自前で cuDNN 8.9 を bundle して動的に load する。 image side の cuDNN は torch 経由の学習スクリプトからは使われない。 動作互換性は保たれる。
- **Ubuntu 22.04 据え置き**: 24.04 への bump は Python 3.11 → 3.12 切り替えや glibc レベル変更を伴うため、 学習スクリプトの動作検証範囲が広がる。 本 PR は OS-package CVE の解消にスコープを絞り、 Ubuntu bump は別 PR で動作確認後に対応する保守的方針。
- **Python 3.11 / torch 2.2.1+cu121 据え置き**: 学習結果の再現性 (RNG seed / numerical stability) に影響を与え得る要素は変更しない。 既存 6lang model / つくよみちゃん FT の再現性を維持。
- **`cudnn8` → `cudnn` tag suffix の変更**: NVIDIA は CUDA 12.4 以降で cuDNN 8 系の image を停止し、 cuDNN 9 系のみ `cudnn` (suffix なし) で提供している (Docker Hub 確認済み: `12.6.3-cudnn-devel-ubuntu22.04` / `12.6.3-cudnn-runtime-ubuntu22.04` 両方 available)。

## Test Plan

- [ ] `docker-build.yml` workflow の `build-python-train` job が green (builder / runtime stage 両方 build 成功)
- [ ] build-time tests (Dockerfile 内 `RUN` の `import torch` / `import pytorch_lightning` / `import piper_train`) が pass
- [ ] `gh workflow run trivy-container-scan.yml --ref chore/bump-python-train-cuda` で Trivy scan を手動実行し、 CVE 件数の削減を確認 (見込み: 約 60% / 約 2,000 件削減)
- [ ] Merge 後の次回 schedule scan (Monday 5:17 UTC) で `piper-plus-python-train` の Trivy alert 数が実測 (受入基準: 削減 ≥ 50%)

## Checklist

- [x] Tests pass locally (build には CUDA GPU 必須のためローカル full build は未実施、 CI で検証)
- [x] No GPL / LGPL dependencies added (base image は NVIDIA EULA + Ubuntu 標準パッケージで既存と同範囲)
- [x] Documentation updated (該当なし — Dockerfile のコメントに bump 理由を inline 記載)

## Related Issues

- Reference: Trivy code-scanning alerts on `piper-plus-python-train` (約 3,469 件)
- 次段階 (別 PR で): Phase 2 で Ubuntu 24.04 + torch 2.3.x への bump を動作検証後に検討

### CODE REVIEW

File: docker/python-train/Dockerfile [hunk 1]
```diff
@@ -2,7 +2,13 @@
...
 
 # Build stage
-FROM nvidia/cuda:12.1.0-cudnn8-devel-ubuntu22.04 AS builder
+# Base: CUDA 12.6.3 + cuDNN 9 + Ubuntu 22.04
+# Note: image side ships cuDNN 9, but torch==2.2.1+cu121 wheel bundles its
+# own cuDNN 8.9 and uses it dynamically at runtime, so cuDNN version mismatch
+# between image and torch does not affect training. Bumped from 12.1.0 to
+# clear Trivy OS-package CVEs (libssl/glibc/libxml2 etc.) accumulated since
+# the 2023-04 base.
+FROM nvidia/cuda:12.6.3-cudnn-devel-ubuntu22.04 AS builder
```
[the remark this answers is not in the document]
Comment (ayutaz) [reply docker/python-train/Dockerfile:22]: ご指摘ありがとうございます。 おっしゃるとおり、 前コメントは cuDNN bundle にのみ言及していて CUDA toolkit (image 12.6 vs torch wheel cu121) の互換性は曖昧でした。 builder stage に以下を追記して clarify しました (commit 91ca5902):

- torch wheel が自前で CUDA 12.1 runtime (libcudart / libcublas / libcufft / libcurand / libcusolver / libcusparse / libcudnn 8.9 / libnccl) を bundle し、 `site-packages/{nvidia,torch/lib}` から動的 load する
- system CUDA 12.6 toolkit (`/usr/local/cuda`) は torch から使われず、 CMake / nvcc 経由の third-party CUDA 消費者 (本 image では現状なし) にのみ供給される
- NVIDIA driver forward compatibility により 12.6 driver で cu121 binary が動作する

これにより 「mixing system CUDA 12.6 libraries with cu121 wheels」 の懸念に対する答え (= torch は system CUDA を mix せず、 自前 bundle で完結する) が Dockerfile から読み取れるようになりました。

File: docker/python-train/Dockerfile [hunk 2]
```diff
@@ -68,7 +74,8 @@ RUN uv pip install "/tmp/src/python/g2p"
 RUN uv pip install "/tmp/src/python[train]"
 
 # Runtime stage
-FROM nvidia/cuda:12.1.0-cudnn8-runtime-ubuntu22.04
+# Base: CUDA 12.6.3 + cuDNN 9 + Ubuntu 22.04 (matches builder stage)
+FROM nvidia/cuda:12.6.3-cudnn-runtime-ubuntu22.04
```
[the remark this answers is not in the document]
Comment (ayutaz) [reply docker/python-train/Dockerfile:94]: ご指摘ありがとうございます。 おっしゃるとおり、 前コメントは cuDNN bundle にのみ言及していて CUDA toolkit (image 12.6 vs torch wheel cu121) の互換性は曖昧でした。 runtime stage に以下を追記して clarify しました (commit 91ca5902):

- torch wheel が自前で CUDA 12.1 runtime (libcudart / libcublas / libcufft / libcurand / libcusolver / libcusparse / libcudnn 8.9 / libnccl) を bundle し、 `site-packages/{nvidia,torch/lib}` から動的 load する
- system CUDA 12.6 toolkit (`/usr/local/cuda`) は torch から使われず、 CMake / nvcc 経由の third-party CUDA 消費者 (本 image では現状なし) にのみ供給される
- NVIDIA driver forward compatibility により 12.6 driver で cu121 binary が動作する

これにより 「mixing system CUDA 12.6 libraries with cu121 wheels」 の懸念に対する答え (= torch は system CUDA を mix せず、 自前 bundle で完結する) が Dockerfile から読み取れるようになりました。
